performance/src/performance/details/is_same_object.cpp

```diff
@@ -0,0 +1,100 @@
+/// @{
+
+is_same_object<measurement>::is_same_object(const measurement& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<measurement>::operator()(const measurement& obj) const {
+
+    return (is_same_scalar(obj.local[0], m_ref.get().local[0], m_unc) &&
+            is_same_scalar(obj.local[1], m_ref.get().local[1], m_unc) &&
+            is_same_scalar(obj.variance[0], m_ref.get().variance[0], m_unc) &&
+            is_same_scalar(obj.variance[1], m_ref.get().variance[1], m_unc));
+}
+
+/// @}
+
+/// @name Implementation for @c traccc::details::is_same_object<seed>
+/// @{
+
+is_same_object<seed>::is_same_object(
+    const spacepoint_container_types::const_view& ref_spacepoints,
+    const spacepoint_container_types::const_view& test_spacepoints,
+    const seed& ref, scalar unc)
+    : m_ref_spacepoints(ref.get_spacepoints(ref_spacepoints)),
+      m_spacepoints(test_spacepoints),
+      m_ref(ref),
+      m_unc(unc) {}
+
+bool is_same_object<seed>::operator()(const seed& obj) const {
+
+    // Extract the spacepoints belonging to the tested seed.
+    std::array<spacepoint, 3> test_spacepoints =
+        obj.get_spacepoints(m_spacepoints);
+
+    // Compare the two seeds.
+    return (is_same_scalar(obj.weight, m_ref.get().weight, m_unc) &&
+            is_same_scalar(obj.z_vertex, m_ref.get().z_vertex, m_unc) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[0],
+                                       m_unc)(test_spacepoints[0]) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[1],
+                                       m_unc)(test_spacepoints[1]) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[2],
+                                       m_unc)(test_spacepoints[2]));
+}
+
+/// @}
+
+/// @name Implementation for @c traccc::details::is_same_object<spacepoint>
+/// @{
+
+is_same_object<spacepoint>::is_same_object(const spacepoint& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<spacepoint>::operator()(const spacepoint& obj) const {
+
+    return (is_same_scalar(obj.x(), m_ref.get().x(), m_unc) &&
+            is_same_scalar(obj.y(), m_ref.get().y(), m_unc) &&
+            is_same_scalar(obj.z(), m_ref.get().z(), m_unc) &&
+            is_same_object<measurement>(m_ref.get().meas, m_unc)(obj.meas));
+}
+
+/// @}
+
+/// @name Implementation for
+///       @c traccc::details::is_same_object<bound_track_parameters>
+/// @{
+
+is_same_object<bound_track_parameters>::is_same_object(
+    const bound_track_parameters& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<bound_track_parameters>::operator()(
+    const bound_track_parameters& obj) const {
+
+    return (is_same_scalar(obj.local()[0], m_ref.get().local()[0], m_unc) &&
+            is_same_scalar(obj.local()[1], m_ref.get().local()[1], m_unc) &&
+            is_same_angle(obj.phi(), m_ref.get().phi(), m_unc) &&
+            is_same_scalar(obj.theta(), m_ref.get().theta(), m_unc) &&
```

Comment:
Shouldn't theta also be treated as an angle?

Reply:
I think he treated it as scalar because phi is in [-pi, pi] but theta is in [0,pi].
I rather wonder if there would be case where phi is defined out of the [-pi,pi] range

Reply:
Indeed. With $\theta$ we don't "flip around". One extreme is in the direction of one of the particle beams, the other extreme is in the direction of the other particle beam. There's no need to worry about the "two extremes being close to each other".

As for why I added a few more `wrap_to_pi` calls in the code than necessarily required: This is mainly out of precaution. While the ATLAS standard is and has been for a long time to define $\phi$ in the [- $\pi$, $\pi$] range, some experiments rather define in in the [0, 2 $\pi$] range. The code that I add in this PR (stolen from https://gitlab.cern.ch/atlas/athena/-/blob/master/Control/CxxUtils/CxxUtils/phihelper.h) tries to give correct results for any combination of values.

And now that I looked at the ATLAS code again, I see that I probably should've gone with

https://gitlab.cern.ch/atlas/athena/-/blob/master/Control/CxxUtils/CxxUtils/phihelper.h#L48-65

to calculate the averages. (I have to admit I didn't think the formula through too much, but if ATLAS does it like that, we might as well do the same in this project.)